plugins/inputs/eventhub/eventhub.go

```diff
@@ -0,0 +1,382 @@
+	PartitionIDs           []string `toml:"partition_ids"`
+	MaxUndeliveredMessages int      `toml:"max_undelivered_messages"`
+	SystemPropertiesPrefix string   `toml:"system_properties_prefix"`
+	EnqTimeTs              bool     `toml:"enq_time_ts"`
+	IotHubEnqTimeTs        bool     `toml:"iot_hub_enq_time_ts"`
+
+	// Azure
+	hub    *eventhub.Hub
+	cancel context.CancelFunc
+
+	// Influx
+	parser parsers.Parser
+
+	// Metrics tracking
+	acc     telegraf.TrackingAccumulator
+	tracker MessageTracker
+	wg      sync.WaitGroup
+}
+
+// MessageTracker is a struct with a lock and list of tracked messages
+type MessageTracker struct {
+	messages map[telegraf.TrackingID][]telegraf.Metric
+	mux      sync.Mutex
+}
+
+// SampleConfig is provided here
+func (*EventHub) SampleConfig() string {
+	return `
+  ## The default behavior is to create a new Event Hub client from environment variables.
+  ## This requires one of the following sets of environment variables to be set:
+  ##
+  ## 1) Expected Environment Variables:
+  ##    - "EVENTHUB_NAMESPACE"
+  ##    - "EVENTHUB_NAME"
+  ##    - "EVENTHUB_CONNECTION_STRING"
+  ##
+  ## 2) Expected Environment Variables:
+  ##    - "EVENTHUB_NAMESPACE"
+  ##    - "EVENTHUB_NAME"
+  ##    - "EVENTHUB_KEY_NAME"
+  ##    - "EVENTHUB_KEY_VALUE"
+
+  ## Uncommenting the option below will create an Event Hub client based solely on the connection string.
+  ## This can either be the associated environment variable or hard coded directly.
+  # connection_string = "$EVENTHUB_CONNECTION_STRING"
+
+  ## Set persistence directory to a valid folder to use a file persister instead of an in-memory persister
+  # persistence_dir = ""
+
+  ## Change the default consumer group
+  # consumer_group = ""
+
+  ## By default the event hub receives all messages present on the broker.
+  ## Alternative modes can be set below. The timestamp should be in RFC3339 format.
+  ## The 3 options below only apply if no valid persister is read from memory or file (e.g. first run).
+  # from_timestamp = ""
+  # starting_offset = ""
+  # latest = true
+
+  ## Set a custom prefetch count for the receiver(s)
+  # prefetch_count = 1000
+
+  ## Add an epoch to the receiver(s)
+  # epoch = 0
+
+  ## Change to set a custom user agent, "telegraf" is used by default
+  # user_agent = "telegraf"
+  
+  ## To consume from a specific partition, set the partition_ids option. 
+  ## An empty array will result in receiving from all partitions.
+  # partition_ids = ["0","1"]
+
+  ## Max undelivered messages
+  # max_undelivered_messages = 1000
+
+  ## Prefix to use for the system properties of Event Hub and IoT Hub messages
+  # system_properties_prefix = "SystemProperties."
+
+  ## Set either option below to true to use a system property as timestamp.
+  ## You have the choice between EnqueuedTime and IoTHubEnqueuedTime.
+  ## It is recommended to use this setting when the data itself has no timestamp.
+  # enq_time_ts = true
+  # iot_hub_enq_time_ts = true
```

Comment:
Are both EnqueuedTime and IoTHubEnqueuedTime set with IoTHub?

Reply:
Yes they are, although one would expect that only `EnqueuedTime` is set when connecting to an Azure Event Hub. I should check if `EnqueuedTime` changes when the IoT Hub message is routed through one or more event hubs.

Reply:
What do you think about removing these options and using iothub time if available, and if not using eventhub time?

Reply:
Nevermind this suggestion, I see now that it won't work well since we need at least one option turn this on/off.

Reply:
Indeed, there could be timestamps within the payload which can be parsed by for example the json parser.